// angular-bootstrap-datetimepicker/esm2015/core/dl-date-time-core.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { DlDateAdapter } from './dl-date-adapter';
import { DlDateAdapterMoment } from './dl-date-adapter-moment';
import { DlDateAdapterNative } from './dl-date-adapter-native';
import { DlDateAdapterNumber } from './dl-date-adapter-number';
import { DlDateAdapterString } from './dl-date-adapter-string';
import { DL_DATE_TIME_DISPLAY_FORMAT, DL_DATE_TIME_DISPLAY_FORMAT_DEFAULT, DL_DATE_TIME_INPUT_FORMATS, DL_DATE_TIME_INPUT_FORMATS_DEFAULT, DL_DATE_TIME_MODEL_FORMAT, DL_DATE_TIME_MODEL_FORMAT_DEFAULT } from './dl-date-time-string-format';
const ɵ0 = DL_DATE_TIME_DISPLAY_FORMAT_DEFAULT, ɵ1 = DL_DATE_TIME_INPUT_FORMATS_DEFAULT, ɵ2 = DL_DATE_TIME_MODEL_FORMAT_DEFAULT;
/**
 * Import this module to supply your own `DateAdapter` provider.
 * @internal
 **/
let DlDateTimeCoreModule = class DlDateTimeCoreModule {
};
DlDateTimeCoreModule = tslib_1.__decorate([
    NgModule({
        providers: [
            { provide: DL_DATE_TIME_DISPLAY_FORMAT, useValue: ɵ0 },
            { provide: DL_DATE_TIME_INPUT_FORMATS, useValue: ɵ1 },
            { provide: DL_DATE_TIME_MODEL_FORMAT, useValue: ɵ2 }
        ]
    })
], DlDateTimeCoreModule);
export { DlDateTimeCoreModule };
/**
 * Import this module to store `milliseconds` in the model.
 * @internal
 */
let DlDateTimeNumberModule = class DlDateTimeNumberModule {
};
DlDateTimeNumberModule = tslib_1.__decorate([
    NgModule({
        imports: [DlDateTimeCoreModule],
        providers: [
            { provide: DlDateAdapter, useClass: DlDateAdapterNumber }
        ],
        exports: [DlDateTimeCoreModule]
    })
], DlDateTimeNumberModule);
export { DlDateTimeNumberModule };
/**
 * Import this module to store a native JavaScript `Date` in the model.
 * @internal
 */
let DlDateTimeDateModule = class DlDateTimeDateModule {
};
DlDateTimeDateModule = tslib_1.__decorate([
    NgModule({
        imports: [DlDateTimeCoreModule],
        providers: [
            { provide: DlDateAdapter, useClass: DlDateAdapterNative }
        ],
    })
], DlDateTimeDateModule);
export { DlDateTimeDateModule };
/**
 * Import this module to store a `moment` in the model.
 * @internal
 */
let DlDateTimeMomentModule = class DlDateTimeMomentModule {
};
DlDateTimeMomentModule = tslib_1.__decorate([
    NgModule({
        imports: [DlDateTimeCoreModule],
        providers: [
            { provide: DlDateAdapter, useClass: DlDateAdapterMoment }
        ],
    })
], DlDateTimeMomentModule);
export { DlDateTimeMomentModule };
const ɵ3 = DL_DATE_TIME_INPUT_FORMATS_DEFAULT, ɵ4 = DL_DATE_TIME_DISPLAY_FORMAT_DEFAULT;
/**
 * Import this module to store a `string` in the model.
 * @internal
 */
let DlDateTimeStringModule = class DlDateTimeStringModule {
};
DlDateTimeStringModule = tslib_1.__decorate([
    NgModule({
        imports: [DlDateTimeCoreModule],
        providers: [
            { provide: DL_DATE_TIME_INPUT_FORMATS, useValue: ɵ3 },
            { provide: DL_DATE_TIME_MODEL_FORMAT, useValue: ɵ4 },
            { provide: DlDateAdapter, useClass: DlDateAdapterString }
        ],
    })
], DlDateTimeStringModule);
export { DlDateTimeStringModule };
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4 };
